### ISSUE TITLE
Lịch sử hình thành TDD

### ISSUE THREAD

**Issue by Jacobvu84** [Jacobvu84/diary#90]
D.D. McCracken (1957): "In order to fully ascertain the accuracy of the answers, it is necessary to have a hand-calculated check case with which to compare the answers which will later be calculated by the machine."

So, to be sure that the machine is giving us the right answers, we need to prepare hand-calculated check cases BEFORE we see the output from the machine.

Edsger W. Dijkstra (1972): "But one should not first make the program and then prove its correctness, because then the requirement of providing the proof would only increase the poor programmer’s burden. On the contrary: the programmer should let correctness proof and program grow hand in hand."

So, we need to first figure out how we'd prove whether the program is correct, and only then to make the program.

Kent Beck: "The original description of TDD was in an ancient book about programming. It said you take the input tape, manually type in the output tape you expect, then program until the actual output tape matches the expected output."

So, it was all about typing in the input the expected output, and about programming until the actual output matches the expected output.

Kent Beck: "Tests should be coupled to the behavior of code and decoupled from the structure of code."

So what's the common theme?

TDD is about formulating expected outputs BEFORE writing the code, to be sure that the program is working correctly, i.e. in accordance with our expectations. It's about proving that our program works correctly.

TDD was always about the expected behavior of the program, NOT about how it's designed.

Notice that, none of the above are saying: is the program designed well? Nope, nowhere.

Tests help us devise the proof that we're satisfied with expected behaviors, NOT that we've designed our system well. When our tests pass, it means our expected behaviors are met. It does NOT mean that our code is well designed or badly designed.

So the tests provide us with the assurance that the system we build behaves as expected. The tests do NOT provide us with assurance whether the system is internally well-designed.

The ONLY architectural/design attribute we achieve through TDD is the testability attribute.

How about all other attributes of quality? Will the system be OOP or FP? Will we write spaghetti code or SOLID code? Will we have a hundred if-else statements or use polymorphism? Whether we'll use arrays or lists or hashmaps? Well, that's up to us.

In conclusion, using TDD provides us assurance that our program is meeting the BEHAVIORAL expectations. It does not dictate our design (only enforces TESTABLITY, nothing else).

- https://agileforall.com/history-of-tdd-as-told-in-quotes
